sopel/irc/backends.py

```diff
@@ -0,0 +1,252 @@
+        AsynchatBackend.__init__(self, bot, **kwargs)
+        self.verify_ssl = verify_ssl
+        self.ssl = None
+        self.ca_certs = ca_certs
+
+    def handle_connect(self):
+        # handle potential TLS connection
+        if self.verify_ssl:
+            self.ssl = ssl.wrap_socket(self.socket,
+                                       do_handshake_on_connect=True,
+                                       suppress_ragged_eofs=True)
+        else:
+            self.ssl = ssl.wrap_socket(self.socket,
+                                       do_handshake_on_connect=True,
+                                       suppress_ragged_eofs=True,
+                                       cert_reqs=ssl.CERT_REQUIRED,
+                                       ca_certs=self.ca_certs)
+            # connect to host specified in config first
+            try:
+                ssl.match_hostname(self.ssl.getpeercert(), self.host)
+            except ssl.CertificateError:
+                # the host in config and certificate don't match
+                LOGGER.error("hostname mismatch between configuration and certificate")
+                # check (via exception) if a CNAME matches as a fallback
+                has_matched = False
+                for hostname in get_cnames(self.host):
+                    try:
+                        ssl.match_hostname(self.ssl.getpeercert(), hostname)
+                        LOGGER.warning(
+                            "using {0} instead of {1} for TLS connection"
+                            .format(hostname, self.host))
+                        has_matched = True
+                        break
+                    except ssl.CertificateError:
+                        pass
+
+                if not has_matched:
+                    # everything is broken
+                    LOGGER.error("invalid certificate, no hostname matches")
+                    # TODO: refactor access to bot's settings
+                    if hasattr(self.bot.settings.core, 'pid_file_path'):
+                        # TODO: refactor to quit properly (no "os._exit")
+                        os.unlink(self.bot.settings.core.pid_file_path)
+                        os._exit(1)
+        self.set_socket(self.ssl)
+        LOGGER.info('Connection accepted by the server...')
+        LOGGER.debug('Starting job scheduler for connection timeout...')
+        self.timeout_scheduler.start()
+        self.bot.on_connect()
+
+    def send(self, data):
+        """Replacement for self.send() during SSL connections."""
+        try:
+            result = self.socket.send(data)
+            return result
+        except ssl.SSLError as why:
+            if why[0] in (asyncore.EWOULDBLOCK, errno.ESRCH):
+                return 0
+            else:
+                raise why
+            return 0
+
+    def recv(self, buffer_size):
+        """Replacement for self.recv() during SSL connections.
+
+        From a (now deleted) blog post by Evan "K7FOS" Fosmark.
```

Comment:
I'd leave the URL even if it errors, in case an archive of it turns up.

https://k7fos.com/2010/09/ssl-support-in-asynchatasync_chat